ray-operator/controllers/ray/batchscheduler/scheduler-plugins/scheduler_plugins.go

```diff
@@ -0,0 +1,154 @@
+		return nil
+	}
+	// we set replica as 1 for the head pod
+	replica := int32(1)
+	for _, workerGroup := range rc.Spec.WorkerGroupSpecs {
+		if workerGroup.Replicas == nil {
+			continue
+		}
+		replica += *workerGroup.Replicas
+	}
+
+	podGroup := &v1alpha1.PodGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: rc.Namespace,
+			Name:      rc.Name,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					Name:       rc.Name,
+					UID:        rc.UID,
+					APIVersion: rc.APIVersion,
+					Kind:       rc.Kind,
+				},
+			},
+		},
+		Spec: v1alpha1.PodGroupSpec{
+			MinMember:    replica,
+			MinResources: utils.CalculateMinResources(rc),
+		},
+	}
+
+	err := k.cli.Create(ctx, podGroup)
+	if errors.IsAlreadyExists(err) {
+		return nil
+	}
+	return err
+}
+
+// AddMetadataToPod adds essential labels and annotations to the Ray pods
+// the scheduler needs these labels and annotations in order to do the scheduling properly
+func (k *KubeScheduler) AddMetadataToPod(_ context.Context, app *rayv1.RayCluster, _ string, pod *corev1.Pod) {
+	// when gang scheduling is enabled, extra labels need to be added to all pods
+	if k.isGangSchedulingEnabled(app) {
+		pod.Labels[KubeSchedulerPodGroupLabelKey] = app.Name
+	}
+}
+
+func (k *KubeScheduler) isGangSchedulingEnabled(app *rayv1.RayCluster) bool {
+	_, exist := app.Labels[utils.RayClusterGangSchedulingEnabled]
+	return exist
+}
+
+func (kf *KubeSchedulerFactory) New(ctx context.Context, c *rest.Config) (schedulerinterface.BatchScheduler, error) {
+	extClient, err := apiextensionsclient.NewForConfig(c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize k8s extension client with error %w", err)
+	}
+
+	podGroupName := "podgroups.scheduling.x-k8s.io"
+	if _, err := extClient.ApiextensionsV1().CustomResourceDefinitions().Get(
+		ctx,
+		podGroupName,
+		metav1.GetOptions{},
+	); err != nil {
+		if _, err := extClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(
+			ctx,
+			podGroupName,
+			metav1.GetOptions{},
+		); err != nil {
+			return nil, fmt.Errorf("podGroup CRD is required to exist in current cluster. error: %w", err)
+		}
+	}
+
+	scheme := runtime.NewScheme()
+	_ = v1alpha1.AddToScheme(scheme)
+	ccache, err := cache.New(c, cache.Options{
+		Scheme: scheme,
+	})
+	if err != nil {
+		return nil, err
+	}
+	go func() {
+		err := ccache.Start(ctx)
+		if err != nil {
+			panic(err)
+		}
+	}()
+	if !ccache.WaitForCacheSync(ctx) {
+		return nil, fmt.Errorf("failed to sync cache")
+	}
+	cli, err := client.New(c, client.Options{
+		Scheme: scheme,
+		Cache: &client.CacheOptions{
+			Reader: ccache,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &KubeScheduler{
+		cli: cli,
+	}, nil
+}
+
+func (kf *KubeSchedulerFactory) AddToScheme(sche *runtime.Scheme) {
+	// No extra scheme needs to be registered
+	_ = v1alpha1.AddToScheme(sche)
+}
+
+func (kf *KubeSchedulerFactory) ConfigureReconciler(b *builder.Builder) *builder.Builder {
+	return b
```

Comment:
should we configure reconciler so that the informer cache has the information of PodGroup?